### ISSUE TITLE
Fix root cause of flaky nightwatch sitemap a11y tests

### ISSUE THREAD

**Issue by billfienberg** [department-of-veterans-affairs/va.gov-team#23331]
## Links

- https://dsva.slack.com/archives/C0MQ281DJ/p1618423474178600
- https://dsva.slack.com/archives/C01RAS1KAQK/p1618320302000900
- [Original PR that enabled parallelization](https://github.com/department-of-veterans-affairs/vets-website/pull/16685)
- [PR to temporarily comment out only first 1/4 of the sitemap tests](https://github.com/department-of-veterans-affairs/vets-website/pull/16773)
- [PR to temporarily comment out all sitemap a11y tests](https://github.com/department-of-veterans-affairs/vets-website/pull/16794)

**Comment by billfienberg** [comment]
>`2 matches found based on name: network nightwatch_default is ambiguous` - https://dsva.slack.com/archives/CBU0KDSB1/p1618524028051500

**Comment by pjhill** [comment]
Daily scheduled a11y sitemap test is not generating the same errors.

**Comment by billfienberg** [comment]
Slack notifications for daily sitemap a11y tests
- [2021-04-05](https://dsva.slack.com/archives/C01RAS1KAQK/p1617630415012100) ✅
- [2021-04-06](https://dsva.slack.com/archives/C01RAS1KAQK/p1617714866012400) ✅
- [2021-04-07](https://dsva.slack.com/archives/C01RAS1KAQK/p1617798711012700) ✅
- [2021-04-08](https://dsva.slack.com/archives/C01RAS1KAQK/p1617885145013000) ✅
- [2021-04-09](https://dsva.slack.com/archives/C01RAS1KAQK/p1617971627000300) ✅
- [2021-04-12](https://dsva.slack.com/archives/C01RAS1KAQK/p1618230779000600) ✅
- [2021-04-13](https://dsva.slack.com/archives/C01RAS1KAQK/p1618320302000900) ❌ 
- [2021-04-14](https://dsva.slack.com/archives/C01RAS1KAQK/p1618403622059100) ✅ 
- [2021-04-15](https://dsva.slack.com/archives/C01RAS1KAQK/p1618490084067600) ✅ 
- [2021-04-16](https://dsva.slack.com/archives/C01RAS1KAQK/p1618576866068000) ✅
- [2021-04-19](https://dsva.slack.com/archives/C01RAS1KAQK/p1618835726072100) ✅
- [2021-04-20](https://dsva.slack.com/archives/C01RAS1KAQK/p1618922235072400) ✅

**Comment by billfienberg** [comment]
```
[2021-04-20T23:15:22.119Z]  src/platform/site-wide/tests/sitemap/sitemap-4.spec   Error retrieving a new session from the selenium server
[2021-04-20T23:15:22.119Z]  src/platform/site-wide/tests/sitemap/sitemap-4.spec   
[2021-04-20T23:15:22.119Z]  src/platform/site-wide/tests/sitemap/sitemap-4.spec   Connection refused! Is selenium server started?
[2021-04-20T23:15:22.119Z]  src/platform/site-wide/tests/sitemap/sitemap-4.spec   Error: connect ECONNREFUSED 172.20.0.3:4444
```

Source: https://dsva.slack.com/archives/C37M86Y8G/p1618962423281800?thread_ts=1618960523.281600&cid=C37M86Y8G

**Comment by billfienberg** [comment]
```
[2021-04-21T18:48:29.951Z]  src/platform/site-wide/tests/sitemap/sitemap-2.spec   
[2021-04-21T18:48:29.951Z]  src/platform/site-wide/tests/sitemap/sitemap-2.spec   Connection refused! Is selenium server started?
[2021-04-21T18:48:29.951Z] 
[2021-04-21T18:48:29.951Z] 
[2021-04-21T18:48:29.951Z]  src/platform/site-wide/tests/sitemap/sitemap-3.spec   Error retrieving a new session from the selenium server
[2021-04-21T18:48:29.951Z] 
[2021-04-21T18:48:29.951Z]  src/platform/site-wide/tests/sitemap/sitemap-2.spec   Error: connect ECONNREFUSED 172.25.0.3:4444
[2021-04-21T18:48:29.951Z]     at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1146:16) {
[2021-04-21T18:48:29.951Z]   errno: -111,
[2021-04-21T18:48:29.951Z]   code: 'ECONNREFUSED',
[2021-04-21T18:48:29.951Z]   syscall: 'connect',
[2021-04-21T18:48:29.951Z]   address: '172.25.0.3',
[2021-04-21T18:48:29.951Z]   port: 4444
[2021-04-21T18:48:29.951Z] }
[2021-04-21T18:48:29.951Z]  src/platform/site-wide/tests/sitemap/sitemap-2.spec   
[2021-04-21T18:48:29.951Z] 
[2021-04-21T18:48:29.951Z]  src/platform/site-wide/tests/sitemap/sitemap-3.spec   
[2021-04-21T18:48:29.951Z]  src/platform/site-wide/tests/sitemap/sitemap-3.spec   Connection refused! Is selenium server started?
[2021-04-21T18:48:29.951Z] 
[2021-04-21T18:48:29.951Z]  src/platform/site-wide/tests/sitemap/sitemap-3.spec   Error: connect ECONNREFUSED 172.25.0.3:4444
[2021-04-21T18:48:29.951Z]     at TCPConnectWrap.afterConnect [as oncomplete] (net.js:1146:16) {
[2021-04-21T18:48:29.951Z]   errno: -111,
[2021-04-21T18:48:29.951Z]   code: 'ECONNREFUSED',
[2021-04-21T18:48:29.951Z]   syscall: 'connect',
[2021-04-21T18:48:29.951Z]   address: '172.25.0.3',
[2021-04-21T18:48:29.951Z]   port: 4444
[2021-04-21T18:48:29.951Z] }
```

http://jenkins.vfs.va.gov/blue/organizations/jenkins/testing%2Fvets-website/detail/master/10922/pipeline

https://dsva.slack.com/archives/C37M86Y8G/p1619030913287700

**Comment by billfienberg** [comment]
I wonder if there's a way to enable merging, but run additional CI steps after. 

Here's a hypothetical example of a staging CI pipeline in my brain:

1. checkout
1. lint/security/unit
1. build
1. integration/e2e tests
1. Set commit to [`success` state](https://docs.github.com/en/rest/reference/repos#statuses) to enable the Merge button
1. deploy to review environment
1. Run sitemap a11y tests

The benefits of this approach could be:
- speed: this would prevent long-running unreliable jobs from blocking teams from merging code
- traceability: we could know if a specific PR broke something
   - When run on a daily schedule, it's more difficult to identify which specific commit broke something
- More data for debugging
   - This would likely expose more bugs in our implementation

**Comment by billfienberg** [comment]
I'm going to try updating the [`start_process` config](https://nightwatchjs.org/gettingstarted/configuration/#selenium-server-settings) from `false` to `true`.

**Comment by billfienberg** [comment]
I just realized [we're using `nightwatch@0.9.21`](https://github.com/department-of-veterans-affairs/vets-website/blob/4ebb0a9a79733ccb4731ebdd16ffded7e8e09175/yarn.lock#L13515-L13529) and the [latest release](https://github.com/nightwatchjs/nightwatch/releases) is `1.6.3`. 

Perhaps we should try upgrading from major version `0` to` major version `1`.

**Comment by billfienberg** [comment]
I spent some time trying to update Nightwatch to see if it would go smoothly. It wasn't too bad, but there are four failing tests left, so I created https://github.com/department-of-veterans-affairs/va.gov-team/issues/23695.

**Comment by billfienberg** [comment]
Other experiments:

- `wrap-nightwatch-accessibility-in-retry` [branch](https://github.com/department-of-veterans-affairs/vets-website/tree/wrap-nightwatch-accessibility-in-retry)
- `nightwatch-accessibility` tests passed 10 times in a row (https://github.com/department-of-veterans-affairs/vets-website/commit/b1991fcc3986fd27fd4cb35c5ae5f7284871283d)
- `update-nightwatch-start-process-config` [branch](https://github.com/department-of-veterans-affairs/vets-website/compare/update-nightwatch-start-process-config)

**Comment by billfienberg** [comment]
Possibly relevant PR: https://github.com/department-of-veterans-affairs/vets-website/pull/7248

**Comment by billfienberg** [comment]
I updated the `start_process` config from `false` to `true` in https://github.com/department-of-veterans-affairs/vets-website/pull/16945, but then the [daily a11y tests failed](https://dsva.slack.com/archives/C01RAS1KAQK/p1619525622074100), so I opened a [PR](https://github.com/department-of-veterans-affairs/vets-website/pull/16958) to change the `start_process` config from `true` back to `false`.

**Comment by billfienberg** [comment]
https://github.com/department-of-veterans-affairs/vets-website/runs/2503684132

```
Error retrieving a new session from the selenium server

Connection refused! Is selenium server started?
{
  value: {
    message: 'Timed out waiting for driver server to start.\n' +
      "Build info: version: '3.141.59', revision: 'e82be7d358', time: '2018-11-14T08:25:53'\n" +
      "System info: host: 'ip-10-247-142-113', ip: '10.247.142.113', os.name: 'Linux', os.arch: 'amd64', os.version: '5.4.0-1028-aws', java.version: '11.0.11'\n" +
      'Driver info: driver.version: unknown',
    error: 'unknown error'
  },
  status: 13
}
```

**Comment by billfienberg** [comment]
Another `Error retrieving a new session from the selenium server` failure

https://dsva.slack.com/archives/C01RAS1KAQK/p1620648618084400

**Comment by billfienberg** [comment]
Ticket to upgrade major version of Nightwatch: #23695

**Comment by billfienberg** [comment]
I have confirmed that it is safe to run `docker-compose down` at the start of the `Integration` stage. In theory, it shouldn't be necessary, but I wanted to see if it was a safe solution to try guaranteeing that the container is down before attempting to start the container. 

https://github.com/department-of-veterans-affairs/vets-website/commit/9c0128d6c4c9812eed5fda4248940f193b5034bd

**Comment by pjhill** [comment]
This is a great record of activity, but this ticket has been OBE. Work to move NW a11y testing to GHA is already underway, and fixes to the quirks of running this test in GHA will be pursued under those tickets.
